File: chapter_4/typescript/recursive_max.ts
const max = (arr: number[], currentMax = 0): number =>
  arr.length === 0
    ? currentMax
    : max(arr.slice(1), arr[0] > currentMax ? arr[0] : currentMax);
  
console.log(max([30, 20, 12, 14, 33, 31, 22]))

File: chapter_4/golang/recursive_max.go
package main

import "fmt"

func max(arr []int) int {

	if len(arr) == 2 {
		if arr[0] > arr[1] {
			return arr[0]
		}
		return arr[1]
	}

	subMax := max(arr[1:])
	if arr[0] > subMax {
		return arr[0]
	}
	return subMax
}

func main() {
	fmt.Print(max([]int{3, 2, 5, 2, 4, 7}))
}
